### DISCUSSION TITLE
코드리뷰 하는 방법

### DISCUSSION THREAD

**Discussion by NEULiee** [Tudy-Project/TUDY#15] [공지]
## TUDY가 코드리뷰 하는 이유
😊  서로의 코드를 점검하고 피드백을 받으며 서로가 성장하는 과정을 가지기 위해서 !

## 코드리뷰 기본 원칙
- 모든 코드는 코드 리뷰를 마친 후에 머지를 해야 합니다.
- 혼자가 아닌 함께 하는 일이기에 우리 모두는 코드 리뷰에 책임감을 갖습니다.
- 자신을 제외한 2명을 리뷰어(assigner)로 지정합니다.
- 1명의 필수 리뷰어가 PR을 승인해야 dev 브랜치에 머지할 수 있습니다.
- PR을 받았으면 하루 안에 답변을 주세요. 답변이 없으면 재촉하세요 !!!
- 리뷰하기에 양이 많다면, 오프라인 리뷰를 요청하세요.
- 자신의 취향을 주장하기 보다는 사실과 데이터를 가지고 의견을 교환해주세요.

### 설계
- 애플리케이션 설계 원칙과 일반적인 프로그래밍 설계 지침에 맞춰서 코드가 작성되어 있는지 확인합니다.

### 유지보수성
- 중요한 값을 하드 코딩을 하고 있지는 않은가요?
- 주석이 코드가 하는 일이 아닌, 코드에 담긴 의도를 설명하고 있나요?
- 코드를 쉽게 이해할 수 있나요?
- 코드만 봐도 작성자의 의도를 이해할 수 있나요?
- 비즈니스 로직을 잘 설명하고 있는지 확인합니다.
- 올바른 변수명, 함수명, 클래스명을 사용했는지 확인합니다.
- 내가 이해할 수 없는 코드는 다른 동료도 이해할 수 없습니다. 내가 이해할 수 없다면 작성자에게 설명을 요구하거나 개선을 부탁합시다!
- 스타일 가이드에 맞게 작성되어 있나요?
- 하나의 함수가 10 라인을 넘어간다면 너무 많은 관심사를 갖고 있는 건 아닌지 유심히 봐주세요.

### 재사용성
- 중복된 코드는 없나요?
- 두 번 이상 반복해서 등장하는 코드는 함수나 클래스로 분리하는 걸 검토하세요.

### 안정성
- 예외 처리를 제대로 하고 있나요?
- 자원을 효율적으로 반환하고 있나요? 예) useEffect의 clean up function에서 제대로 초기화를 했는지

### 확장성
- 새로운 기능을 추가하기 쉽게 작성되어 있나요?
- 함수나 클래스가 두 개 이상의 관심사를 갖고 있지는 않은지 확인해주세요.

### 테스트 
- 모든 코드는 테스트 하기가 쉬워야 합니다.
- 큰 함수를 작은 단위로 나눠주세요.
- 클래스 간에 모킹하기 쉬운 인터페이스로 연결되어 있는지 확인해주세요.
- 테스트가 작성되어 있는지 반드시 확인을 합니다.
- 테스트가 작성되어 있다면 [테스트 작성 가이드](https://github.com/meshkorea/front-end-engineering/blob/main/tests/index.md)를 참고해서 개선할 부분이 있는지를 확인합니다.

## 리뷰중인 코드 보는 방법

- 다음과 같은 단계로 리뷰를 해보세요!
  1. 전체 코드를 훑어보기
     - 코드 전체에서 일어나는 변경을 가볍게 훑어보세요.
     - 의미가 없는 변경점이나 변경이 일어나서는 안 되는 곳이 있는지 확인합니다.
  2. PR의 주안점 확인
     - 리뷰를 받는 사람이 제시한 주요 변경사항을 확인합니다.
     - 주안점이 제시되어 있지 않다면 PR 작성자에게 설명을 요구합니다.
     - 구조를 변경한 PR은 빠르게 피드백을 전달하고 동료에게 전파를 해야합니다.
  3. 놓치는 파일이 없도록 모든 부분을 꼼꼼하게 확인합니다.
- 잘한 점이 있다면 칭찬해 주세요!
- 리뷰를 해야 할 코드의 양이 너무 많다면? 오프라인으로 리뷰를 요청하세요.

## 언제 코드 리뷰를 하면 좋을까?

- 코드 리뷰는 요청을 받는 즉시 하는 걸 원칙으로 합니다.
- PR을 받았으면 하루 안에는 답변을 주세요. 상대방의 답변이 없으면 재촉하세요.

<br><br>

[출처: meshkorea github](https://github.com/meshkorea/front-end-engineering/blob/main/conventions/code-review/index.md)

**Comment by realswxx** [comment]
아주 좋습니닼ㅋ
